### PR TITLE
Add some unit tests & improve logging

### PR DESCRIPTION
**Pull request by chakri-nelluri**
Add some unit tests & improve logging

### CODE REVIEW

File: pkg/csi-common/driver_test.go [hunk 1]
```diff
@@ -0,0 +1,165 @@
+package csicommon
```
Comment (rootfs) [inline pkg/csi-common/driver_test.go]: license

File: pkg/csi-common/identityserver-default_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package csicommon
```
Comment (rootfs) [inline pkg/csi-common/identityserver-default_test.go]: ditto

File: pkg/csi-common/nodeserver-default_test.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
+package csicommon
```
Comment (rootfs) [inline pkg/csi-common/nodeserver-default_test.go]: ditto

File: pkg/csi-common/driver_test.go [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+		Minor: 3,
+		Patch: 0,
+	}
+	err = driver.CheckVersion(&v)
+	s, ok = status.FromError(err)
+	assert.True(t, ok)
+	assert.Equal(t, s.Code(), codes.InvalidArgument)
+}
+
+func TestGetVolumeCapabilityAccessModes(t *testing.T) {
```
Comment (rootfs) [inline pkg/csi-common/driver_test.go]: what about access type (mount, block)?

File: pkg/csi-common/server.go [hunk 5]
```diff
@@ -19,14 +19,58 @@ package csicommon
...
+	go s.serve(endpoint, ids, cs, ns)
+
+	return
+}
+
+func (s *nonBlockingGRPCServer) Wait() {
+	s.wg.Wait()
+}
+
+func (s *nonBlockingGRPCServer) Stop() {
```
Comment (rootfs) [inline pkg/csi-common/server.go]: When to stop?

File: pkg/csi-common/server.go [hunk 6]
```diff
@@ -60,4 +108,5 @@ func Serve(endpoint string, ids csi.IdentityServer, cs csi.ControllerServer, ns
 	glog.Infof("Listening for connections on path: %#v", listener.Addr())
```
Comment (rootfs) [inline pkg/csi-common/server.go]: on path -> on address

File: pkg/csi-common/utils.go [hunk 7]
```diff
@@ -21,12 +21,17 @@ import (
...
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
 )
 
 func ParseEndpoint(ep string) (string, string, error) {
 	if strings.HasPrefix(strings.ToLower(ep), "unix://") || strings.HasPrefix(strings.ToLower(ep), "tcp://") {
 		s := strings.SplitN(ep, "://", 2)
-		return s[0], s[1], nil
+		if s[1] != "" {
```
Comment (rootfs) [inline pkg/csi-common/utils.go]: validate len(s) too

File: pkg/csi-common/utils_test.go [hunk 8]
```diff
@@ -0,0 +1,87 @@
+package csicommon
```
Comment (rootfs) [inline pkg/csi-common/utils_test.go]: license

File: pkg/csi-common/utils_test.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+	_, _, err = ParseEndpoint("unix:/fake.sock/")
+	assert.NotNil(t, err)
+
+	_, _, err = ParseEndpoint("fake.sock")
+	assert.NotNil(t, err)
+
+	_, _, err = ParseEndpoint("unix://")
+	assert.NotNil(t, err)
+
+	_, _, err = ParseEndpoint("")
```
Comment (rootfs) [inline pkg/csi-common/utils_test.go]: test `://`

File: pkg/csi-common/driver_test.go (hunk 1 above)
Comment (chakri-nelluri) [reply pkg/csi-common/driver_test.go]: Done

File: pkg/csi-common/identityserver-default_test.go (hunk 2 above)
Comment (chakri-nelluri) [reply pkg/csi-common/identityserver-default_test.go]: Done

File: pkg/csi-common/nodeserver-default_test.go (hunk 3 above)
Comment (chakri-nelluri) [reply pkg/csi-common/nodeserver-default_test.go]: Done

File: pkg/csi-common/driver_test.go (hunk 4 above)
Comment (chakri-nelluri) [reply pkg/csi-common/driver_test.go]: That is VolumeCapability. As of now, we don't expose VolumeCapability in the driver base. Let me see if I can abstract it out too. Will be next PR though :)

File: pkg/csi-common/server.go (hunk 5 above)
Comment (chakri-nelluri) [reply pkg/csi-common/server.go]: This is following the pattern which @lpabon is using for testing. Start a driver, and stop a driver and then start again etc.
Today, we don't trap any signals, after we add signal handlers, stop() will be called from signal handler.

File: pkg/csi-common/server.go (hunk 6 above)
Comment (chakri-nelluri) [reply pkg/csi-common/server.go]: Conscience programmed to unix.sock :)

File: pkg/csi-common/utils.go (hunk 7 above)
Comment (chakri-nelluri) [reply pkg/csi-common/utils.go]: There is a lot of validation code missing. CSI specs defined length limits & restrictions on various parameters. Will add it, as part of it.

File: pkg/csi-common/utils_test.go (hunk 8 above)
Comment (chakri-nelluri) [reply pkg/csi-common/utils_test.go]: Done

File: pkg/csi-common/utils_test.go (hunk 9 above)
Comment (chakri-nelluri) [reply pkg/csi-common/utils_test.go]: Done.

**Comment by chakri-nelluri** [comment]
Thanks @rootfs. Addressed review comments. PTAL when you get a chance.
Let me know if everything looks good. I will rebase before merge.

File: pkg/cinder/mount/mount.go [hunk 10]
```diff
@@ -25,13 +25,13 @@ import (
 	"k8s.io/kubernetes/pkg/util/mount"
 	"k8s.io/kubernetes/pkg/volume/util"
 	utilexec "k8s.io/utils/exec"
-
```
Comment (rootfs) [inline pkg/cinder/mount/mount.go]: we need to follow up with gofmt checker in CI

**Comment by rootfs** [comment]
lgtm, feel free to merge

File: pkg/cinder/mount/mount.go (hunk 10 above)
Comment (chakri-nelluri) [reply pkg/cinder/mount/mount.go]: +1 Yes. I am doing it manually now.

**Comment by edisonxiang** [comment]
@chakri-nelluri Great Job.

**Pull request merged**